File: class_file.py
import struct
import sys
from binascii import hexlify

def parse_row(s):
    a, b, c, d = s.strip().split(',')
    return int(a), c, int(d)

def parse_opcode_table(buf):
    rows = [parse_row(i) for i in buf.strip().split('\n')]
    return {opcode: (name, args) for opcode, name, args in rows}

with open('opcodes.csv', 'r') as f:
    opcode_table = parse_opcode_table(f.read())

def size_to_format(size):
    if size == 4:
        return ">I"
    elif size == 2:
        return ">H"
    elif size == 1:
        return ">B"
    else:
        assert False, size

def field(buf, size):
    format = size_to_format(size)
    value, = struct.unpack(format, buf[0:size])
    return buf[size:], value

constant_lookup = {
    7: ("CONSTANT_Class", [("name_index", 2)]),
    9: ("CONSTANT_Fieldref", [("class_index", 2), ("name_and_type_index", 2)]),
    10: ("CONSTANT_Methodref", [("class_index", 2), ("name_and_type_index", 2)]),
    11: ("CONSTANT_InterfaceMethodref", [("class_index", 2), ("name_and_type_index", 2)]),
    8: ("CONSTANT_String", [("string_index", 2)]),
    3: ("CONSTANT_Integer", [("bytes", 4)]),
    4: ("CONSTANT_Float", [("bytes", 4)]),
    5: ("CONSTANT_Long", [("high_bytes", 4), ("low_bytes", 4)]),
    6: ("CONSTANT_Double", [("high_bytes", 4), ("low_bytes", 4)]),
    12: ("CONSTANT_NameAndType", [("name_index", 2), ("descriptor_index", 2)]),
    1: ("CONSTANT_Utf8", [("length", 2), ("bytes", 0)]),
    15: ("CONSTANT_MethodHandle", [("reference_kind", 2), ("reference_index", 2)]),
    16: ("CONSTANT_MethodType", [("descriptor_index", 2)]),
    18: ("CONSTANT_InvokeDynamic", [("bootstrap_method_attr_index", 2), ("name_and_type_index", 2)]),
}


def parse_cp_info(buf):
    buf, tag = field(buf, 1)
    assert tag in constant_lookup, tag
    name, fields = constant_lookup[tag]
    d = {}
    last_value = None
    for field_name, field_size in fields:
        if field_size == 0:
            value = bytes(buf[0:last_value])
            buf = buf[last_value:]
        else:
            buf, value = field(buf, field_size)
            last_value = value
        d[field_name] = value
    return buf, (name, d)

def parse_attribute_info(buf):
    buf, attribute_name_index = field(buf, 2)
    buf, attribute_length = field(buf, 4)
    info = bytes(buf[0:attribute_length])
    buf = buf[attribute_length:]
    return buf, (attribute_name_index, attribute_length, info)

def parse_field_info(buf):
    buf, access_flags = field(buf, 2)
    buf, name_index = field(buf, 2)
    buf, descriptor_index = field(buf, 2)
    buf, attributes_count = field(buf, 2)
    attributes = []
    for i in range(attributes_count):
        buf, attribute = parse_attribute_info(buf)
        attributes.append(attribute)
    return buf, (access_flags, name_index, descriptor_index, attributes_count, attributes)

def parse_exception_table(buf, indent):
    exception_table = [
        ("start_pc", 2),
        ("end_pc", 2),
        ("handler_pc", 2),
        ("catch_type", 2),
    ]
    for name, size in exception_table:
        buf, value = field(buf, size)
        print(indent, name, value)
    return buf

def dump_opcodes(buf, indent):
    ix = 0
    while len(buf) > 0:
        op = buf[0]
        name, arg_length = opcode_table[op]
        buf = buf[1:]
        args = list(buf[0:arg_length])
        print(indent, f"{ix:> 3}", name, args)
        ix += 1 + arg_length
        buf = buf[arg_length:]

def print_code_info(buf, indent, constant_pool):
    code_attribute = [
        ("max_stack", 2),
        ("max_locals", 2),
        ("code_length", 4),
        ("code", 0),
        ("exception_table_length", 2),
        ("exception_table", 0),
        ("attributes_count", 2),
        ("attributes", 0)
    ]
    last_value = None
    for name, size in code_attribute:
        if name == "code":
            print(indent, "code:")
            size = last_value
            dump_opcodes(buf[0:last_value], indent + "  ")
            buf = buf[last_value:]
        elif name == "exception_table":
            print(indent, "exception_table:")
            for i in range(last_value):
                buf = parse_exception_table(buf, indent + "  ")
        elif name == "attributes":
            print(indent, "attributes:")
            for i in range(last_value):
                buf, attribute_info = parse_attribute_info(buf)
                print(indent + "  ", f"attribute {i}:")
                attribute_name_index, attribute_length, info = attribute_info

                constant_type, constant = constant_pool[attribute_name_index - 1]
                assert constant_type == 'CONSTANT_Utf8', constant_type
                attribute_name_bytes = constant["bytes"]

                print(indent + "    ", "attribute_name_index", attribute_name_index, attribute_name_bytes)
                print(indent + "    ", "attribute_length", attribute_length)
                print(indent + "    ", "info", info)
        else:
            buf, value = field(buf, size)
            last_value = value
            print(indent, name, value)
    assert len(buf) == 0

def parse_class(buf):
    buf, magic = field(buf, 4)
    buf, minor_version = field(buf, 2)
    buf, major_version = field(buf, 2)
    buf, constant_pool_count = field(buf, 2)

    print("magic", hex(magic))
    print("minor_version", minor_version)
    print("major_version", major_version)
    print("constant_pool_count", constant_pool_count)

    print("constant_pool:")
    constant_pool = []
    for i in range(constant_pool_count - 1):
        buf, cp_info = parse_cp_info(buf)
        constant_pool.append(cp_info)
        print(i + 1, cp_info)

    buf, access_flags = field(buf, 2)
    buf, this_class = field(buf, 2)
    buf, super_class = field(buf, 2)
    buf, interfaces_count = field(buf, 2)

    print("access_flags", hex(access_flags))
    print("this_class", this_class)
    print("super_class", super_class)
    print("interfaces_count", interfaces_count)

    print("interfaces:")
    for i in range(interfaces_count):
        buf, interface = field(buf, 2)
        print(i, interface)

    buf, fields_count = field(buf, 2)
    print("fields_count", fields_count)
    for i in range(fields_count):
        buf, field_info = parse_field_info(buf)
        print(i, field_info)

    buf, methods_count = field(buf, 2)
    print("methods_count", methods_count)
    print("methods:")
    for i in range(methods_count):
        buf, method_info = parse_field_info(buf)
        print(f"  method {i}:")
        access_flags, name_index, descriptor_index, attributes_count, attributes = method_info
        print("    access_flags", hex(access_flags))

        constant_type, constant = constant_pool[name_index - 1]
        assert constant_type == 'CONSTANT_Utf8', constant_type
        name_bytes = constant["bytes"]

        print("    name_index", name_index, name_bytes)

        constant_type, constant = constant_pool[descriptor_index - 1]
        assert constant_type == 'CONSTANT_Utf8', constant_type
        descriptor_bytes = constant["bytes"]

        print("    descriptor_index", descriptor_index, descriptor_bytes)
        print("    attributes_count", attributes_count)
        print("    attributes:")
        for j in range(attributes_count):
            print(f"    attribute {j}:")
            attribute_name_index, attribute_length, info = attributes[j]

            constant_type, constant = constant_pool[attribute_name_index - 1]
            assert constant_type == 'CONSTANT_Utf8', constant_type
            attribute_name_bytes = constant["bytes"]

            print("      attribute_name_index", attribute_name_index, attribute_name_bytes)
            print("      attribute_length", attribute_length)
            #print("      info", info)
            if attribute_name_bytes == b'Code':
                print_code_info(info, "        ", constant_pool)

    buf, attributes_count = field(buf, 2)
    print("attributes_count", attributes_count)
    print("attributes:")
    for i in range(attributes_count):
        buf, attribute_info = parse_attribute_info(buf)
        print(i, attribute_info)

    assert len(buf) == 0, bytes(buf)

filename = sys.argv[1]

with open(filename, 'rb') as f:
    buf = f.read()

parse_class(memoryview(buf))

File: c/class_file.c
#include "assert.h"
#include "class_file.h"
#include "malloc.h"
#include "bytes.h"

struct attribute_info * attribute_info_parse(const uint8_t ** buf, int attribute_count, struct constant * constant_pool);

void * attribute_info_parse_info(const uint8_t * buf, struct attribute_info * attribute, struct constant * constant_pool)
{
  struct constant * attribute_name = &constant_pool[attribute->attribute_name_index - 1];
#ifdef DEBUG
  assert(attribute_name->tag == CONSTANT_Utf8);
#endif

  if (constant_equal(attribute_name, "ConstantValue")) {
    struct ConstantValue_attribute * constantvalue = malloc_class_arena((sizeof (struct ConstantValue_attribute)));
    constantvalue->constantvalue_index = parse_u2(&buf);
    return constantvalue;
  } else if (constant_equal(attribute_name, "Code")) {
    // parse Code
    struct Code_attribute * code = malloc_class_arena((sizeof (struct Code_attribute)));

    code->max_stack = parse_u2(&buf);
    code->max_locals = parse_u2(&buf);
    code->code_length = parse_u4(&buf);

    code->code = parse_bytes(&buf, code->code_length);
    code->exception_table_length = parse_u2(&buf);
    uint32_t exception_table_size = (sizeof (struct exception_table_entry)) * code->exception_table_length;
    code->exception_table = malloc_class_arena(exception_table_size);
    for (int i = 0; i < code->exception_table_length; i++) {
      // parse exception_table_entry
      code->exception_table[i].start_pc = parse_u2(&buf);
      code->exception_table[i].end_pc = parse_u2(&buf);
      code->exception_table[i].handler_pc = parse_u2(&buf);
      code->exception_table[i].catch_type = parse_u2(&buf);
    }
    code->attributes_count = parse_u2(&buf);

    code->attributes = attribute_info_parse(&buf, code->attributes_count, constant_pool);

    return code;
  } else if (constant_equal(attribute_name, "BootstrapMethods")) {
    // parse BootstrapMethods
    struct BootstrapMethods_attribute * bootstrap_methods = malloc_class_arena((sizeof (struct BootstrapMethods_attribute)));

    bootstrap_methods->num_bootstrap_methods = parse_u2(&buf);

    uint32_t bootstrap_methods_size = (sizeof (struct bootstrap_method)) * bootstrap_methods->num_bootstrap_methods;
    bootstrap_methods->bootstrap_methods = malloc_class_arena(bootstrap_methods_size);

    for (int i = 0; i < bootstrap_methods->num_bootstrap_methods; i++) {
      bootstrap_methods->bootstrap_methods[i].bootstrap_method_ref = parse_u2(&buf);
      bootstrap_methods->bootstrap_methods[i].num_bootstrap_arguments = parse_u2(&buf);

      uint32_t bootstrap_arguments_size = (sizeof (u2)) * bootstrap_methods->bootstrap_methods[i].num_bootstrap_arguments;
      bootstrap_methods->bootstrap_methods[i].bootstrap_arguments = malloc_class_arena(bootstrap_arguments_size);
      for (int j = 0; j < bootstrap_methods->bootstrap_methods[i].num_bootstrap_arguments; j++) {
        bootstrap_methods->bootstrap_methods[i].bootstrap_arguments[j] = parse_u2(&buf);
      }
    }

    return bootstrap_methods;
  } else if (constant_equal(attribute_name, "NestHost")) {
    struct NestHost_attribute * nesthost = malloc_class_arena((sizeof (struct NestHost_attribute)));
    nesthost->host_class_index = parse_u2(&buf);

    return nesthost;
  } else if (constant_equal(attribute_name, "NestMembers")) {
    struct NestMembers_attribute * nestmembers = malloc_class_arena((sizeof (struct NestMembers_attribute)));

    nestmembers->number_of_classes = parse_u2(&buf);
    uint32_t classes_size = (sizeof (u2)) * nestmembers->number_of_classes;
    nestmembers->classes = malloc_class_arena(classes_size);

    for (int i = 0; i < nestmembers->number_of_classes; i++) {
      nestmembers->classes[i] = parse_u2(&buf);
    }

    return nestmembers;
  } else if (constant_equal(attribute_name, "LineNumberTable")) {
    struct LineNumberTable_attribute * line_number_table = malloc_class_arena((sizeof (struct LineNumberTable_attribute)));
    line_number_table->line_number_table_length = parse_u2(&buf);

    uint32_t line_number_table_size = (sizeof (struct line_number_table)) * line_number_table->line_number_table_length;
    line_number_table->line_number_table = malloc_class_arena(line_number_table_size);

    for (int i = 0; i < line_number_table->line_number_table_length; i++) {
      line_number_table->line_number_table[i].start_pc = parse_u2(&buf);
      line_number_table->line_number_table[i].line_number = parse_u2(&buf);
    }
    return line_number_table;
  } else if (constant_equal(attribute_name, "Exceptions")) {
    struct Exceptions_attribute * exceptions = malloc_class_arena((sizeof (struct Exceptions_attribute)));
    exceptions->number_of_exceptions = parse_u2(&buf);
    uint32_t exceptions_size = (sizeof (u2)) * exceptions->number_of_exceptions;
    exceptions->exception_index_table = malloc_class_arena(exceptions_size);
    for (int i = 0; i < exceptions->number_of_exceptions; i++) {
      exceptions->exception_index_table[i] = parse_u2(&buf);
    }
    return exceptions;
  } else {
    return nullptr;
  }

  assert(false);
}

struct attribute_info * attribute_info_parse(const uint8_t ** buf, int attribute_count, struct constant * constant_pool)
{
  uint32_t attributes_size = (sizeof (struct attribute_info)) * attribute_count;
  struct attribute_info * attributes = malloc_class_arena(attributes_size);
  for (int i = 0; i < attribute_count; i++) {
    attributes[i].attribute_name_index = parse_u2(buf);
    attributes[i].attribute_length = parse_u4(buf);

    attributes[i].info = attribute_info_parse_info(*buf, &attributes[i], constant_pool);

    (*buf) += attributes[i].attribute_length;
  }
  return attributes;
}

struct class_file * class_file_parse(const uint8_t * buf)
{
  struct class_file * class_file = malloc_class_arena((sizeof (struct class_file)));

  class_file->magic = parse_u4(&buf);
  class_file->minor_version = parse_u2(&buf);
  class_file->major_version = parse_u2(&buf);
  class_file->constant_pool_count = parse_u2(&buf);

  uint32_t constant_pool_size = (sizeof (struct constant)) * (class_file->constant_pool_count - 1);
  class_file->constant_pool = malloc_class_arena(constant_pool_size);

  // parse constants
  for (int i = 0; i < class_file->constant_pool_count - 1; i++) {
    struct constant * constant = &class_file->constant_pool[i];
    u1 tag = parse_u1(&buf);
    #ifdef DEBUG
    constant->tag = tag;
    #endif
    switch (tag) {
    case CONSTANT_Class:
      constant->class.name_index = parse_u2(&buf);
      break;
    case CONSTANT_Fieldref:             /* fall through */;
    case CONSTANT_Methodref:            /* fall through */;
    case CONSTANT_InterfaceMethodref:
      constant->fieldref.class_index = parse_u2(&buf);
      constant->fieldref.name_and_type_index = parse_u2(&buf);
      break;
    case CONSTANT_String:
      constant->string.string_index = parse_u2(&buf);
      break;
    case CONSTANT_Integer:              /* fall through */;
    case CONSTANT_Float:
      constant->integer.bytes = parse_u4(&buf);
      break;
    case CONSTANT_Long:                 /* fall through */;
    case CONSTANT_Double:
      {
        uint64_t high_bytes = parse_u4(&buf);
        uint64_t low_bytes = parse_u4(&buf);
        constant->_long.bytes = high_bytes << 32 | low_bytes;
        // long and double use two constant pool entries
        class_file->constant_pool[i + 1].tag = 0;
        i += 1;
      }
      break;
    case CONSTANT_NameAndType:
      constant->nameandtype.name_index = parse_u2(&buf);
      constant->nameandtype.descriptor_index = parse_u2(&buf);
      break;
    case CONSTANT_Utf8:
      constant->utf8.length = parse_u2(&buf);
      constant->utf8.bytes = parse_bytes(&buf, constant->utf8.length);
      break;
    case CONSTANT_MethodHandle:
      constant->methodhandle.reference_kind = parse_u1(&buf);
      constant->methodhandle.reference_index = parse_u2(&buf);
      break;
    case CONSTANT_MethodType:
      constant->methodtype.descriptor_index = parse_u2(&buf);
      break;
    case CONSTANT_Dynamic:              /* fall through */;
    case CONSTANT_InvokeDynamic:
      constant->dynamic.bootstrap_method_attr_index = parse_u2(&buf);
      constant->dynamic.name_and_type_index = parse_u2(&buf);
      break;
    case CONSTANT_Module:               /* fall through */;
    case CONSTANT_Package:
      constant->module.name_index = parse_u2(&buf);
      break;
    default:
      assert(false);
      break;
    }
  }

  class_file->access_flags = parse_u2(&buf);
  class_file->this_class = parse_u2(&buf);
  class_file->super_class = parse_u2(&buf);
  class_file->interfaces_count = parse_u2(&buf);

  // parse interfaces
  uint32_t interfaces_size = (sizeof (class_file->interfaces[0])) * class_file->interfaces_count;
  class_file->interfaces = malloc_class_arena(interfaces_size);
  for (int i = 0; i < class_file->interfaces_count; i++)
    class_file->interfaces[i] = parse_u2(&buf);

  // parse fields
  class_file->fields_count = parse_u2(&buf);
  uint32_t fields_size = (sizeof (struct field_info)) * class_file->fields_count;
  class_file->fields = malloc_class_arena(fields_size);
  for (int i = 0; i < class_file->fields_count; i++) {
    class_file->fields[i].access_flags = parse_u2(&buf);
    class_file->fields[i].name_index = parse_u2(&buf);
    class_file->fields[i].descriptor_index = parse_u2(&buf);
    class_file->fields[i].attributes_count = parse_u2(&buf);

    // parse attributes
    class_file->fields[i].attributes = attribute_info_parse(&buf, class_file->fields[i].attributes_count, class_file->constant_pool);
  }


  // parse methods
  class_file->methods_count = parse_u2(&buf);
  uint32_t methods_size = (sizeof (struct method_info)) * class_file->methods_count;
  class_file->methods = malloc_class_arena(methods_size);
  for (int i = 0; i < class_file->methods_count; i++) {
    class_file->methods[i].access_flags = parse_u2(&buf);
    class_file->methods[i].name_index = parse_u2(&buf);
    class_file->methods[i].descriptor_index = parse_u2(&buf);
    class_file->methods[i].attributes_count = parse_u2(&buf);

    // parse attributes
    class_file->methods[i].attributes = attribute_info_parse(&buf, class_file->methods[i].attributes_count, class_file->constant_pool);
  }

  // parse attributes
  class_file->attributes_count = parse_u2(&buf);
  class_file->attributes = attribute_info_parse(&buf, class_file->attributes_count, class_file->constant_pool);

  return class_file;
}
